### PR TITLE
implemented the remaining builtin functions

### PR DESCRIPTION
**Pull request by VikramKuberappa**

### CODE REVIEW

File: python_builtin_functions.py [hunk 1]
```diff
@@ -4,86 +4,170 @@ def myabs(number):
...
+    for item in iterable:
+        if item:
+            return True
+    return False
 
 
 def mylen(s):
     """ Implementation of built-in `len` function """
-    pass
+    i=0
```
Comment (sumanth-lingappa) [inline python_builtin_functions.py]: You can rename `i` as `count` or `length`

File: python_builtin_functions.py [hunk 2]
```diff
@@ -4,86 +4,170 @@ def myabs(number):
...
+    for _ in s:
+        i+=1
+    return i
 
 
 def mysum(iterable):
     """ Implementation of built-in `sum` function """
-    pass
+    i=0
```
Comment (sumanth-lingappa) [inline python_builtin_functions.py]: You can do proper naming instead of `i`

File: python_builtin_functions.py [hunk 3]
```diff
@@ -4,86 +4,170 @@ def myabs(number):
...
 
 
 def mymax(iterable):
     """ Implementation of built-in `max` function """
-    pass
+    i=0
+    for x in iterable:
+        if x>i:
+            i=x
+        else:
```
Comment (sumanth-lingappa) [inline python_builtin_functions.py]: `else` block is not required as there is only one `continue` statement and no other statements are present

File: python_builtin_functions.py [hunk 4]
```diff
@@ -4,86 +4,170 @@ def myabs(number):
...
+    i=0
+    for x in iterable:
+        i+=x
+    return i
 
 
 def mymax(iterable):
     """ Implementation of built-in `max` function """
-    pass
+    i=0
```
Comment (sumanth-lingappa) [inline python_builtin_functions.py]: This initialisation of `i` to zero doesn't work for a list containing only negative integers..

Instead, initialise `i` to the first element , thereby assuming the max element as first element and comparing the max element with rest of the list items.

File: python_builtin_functions.py [hunk 5]
```diff
@@ -4,86 +4,170 @@ def myabs(number):
...
+            i=x
+        else:
+            continue
+    return i
 
 
 def mymin(iterable):
     """ Implementation of built-in `min` function """
-    pass
+    i=1000**1000
```
Comment (sumanth-lingappa) [inline python_builtin_functions.py]: This doesn't work for a list whole all the elements are greater than `1000*1000`.

Instead like `mymax` function, initialize `i` to the first element of `iterable`

File: python_builtin_functions.py [hunk 6]
```diff
@@ -4,86 +4,170 @@ def myabs(number):
...
+    for x in iterable:
+        if x<i:
+            i=x
+        else:
```
Comment (sumanth-lingappa) [inline python_builtin_functions.py]: `else` block not required as stated above

File: python_builtin_functions.py [hunk 7]
```diff
@@ -4,86 +4,170 @@ def myabs(number):
...
+            continue
+    return i
 
 
 def mydivmod(a, b):
     """ Implementation of built-in `divmod` function """
-    pass
+    q=a//b
```
Comment (sumanth-lingappa) [inline python_builtin_functions.py]: Further refactoring ...

You can use a one liner as below..

`return (a//b, a%b)`

File: python_builtin_functions.py [hunk 8]
```diff
@@ -4,86 +4,170 @@ def myabs(number):
...
+    r=a%b
+    return (q,r)
 
 
 def mybool(x):
     """ Implementation of built-in `bool` function """
-    pass
+    if x:
```
Comment (sumanth-lingappa) [inline python_builtin_functions.py]: You can use one liner as below..

`return True if x else False`

File: python_builtin_functions.py [hunk 9]
```diff
@@ -4,86 +4,170 @@ def myabs(number):
...
+        return True
+    return False
 
 
 def mypow(base, exp):
     """ Implementation of built-in `pow` function """
-    pass
+    i=base**exp
```
Comment (sumanth-lingappa) [inline python_builtin_functions.py]: Again, variable `i` is not required..

You can directly

`return base**exp`

File: python_builtin_functions.py [hunk 10]
```diff
@@ -4,86 +4,170 @@ def myabs(number):
...
+        x=x//2
+        lst.insert(1,str(r))
+    return ''.join(lst)
 
 
 def myhex(x):
     """ Implementation of built-in `hex` function """
-    pass
+    lst=['0x']
+    while int(x>0):
```
Comment (sumanth-lingappa) [inline python_builtin_functions.py]: Why converting `x` to `int` again?

File: python_builtin_functions.py [hunk 11]
```diff
@@ -4,86 +4,170 @@ def myabs(number):
...
-    pass
+    lst=['0b']
+    while x>0:
+        r=x%2
+        x=x//2
+        lst.insert(1,str(r))
+    return ''.join(lst)
 
 
 def myhex(x):
```
Comment (sumanth-lingappa) [inline python_builtin_functions.py:106]: Check if this works for `myhex(0)`

File: python_builtin_functions.py [hunk 12]
```diff
@@ -4,86 +4,170 @@ def myabs(number):
...
-    pass
+    lst=[]
+    tup=()
+    for x in range (0,len(iterable)):
+        tup=(x,iterable[x])
+        lst.append(tup)
+    return lst
 
 
 def mybin(x):
```
Comment (sumanth-lingappa) [inline python_builtin_functions.py:93]: Check of this works for `mybin(0)`

File: python_builtin_functions.py [hunk 13]
```diff
@@ -4,86 +4,170 @@ def myabs(number):
...
+            r='e'
+        elif r==15:
+            r='f'
+        else:
+            pass
+        lst.insert(1,str(r))
+    return ''.join(lst)
 
 
 def myoct(x):
```
Comment (sumanth-lingappa) [inline python_builtin_functions.py:133]: Again, check if this works for `myoct(0)`

File: python_builtin_functions.py [hunk 14]
```diff
@@ -4,86 +4,170 @@ def myabs(number):
...
     """ Implementation of built-in `oct` function """
-    pass
+    lst=['0o']
+    while int(x>0):
```
Comment (sumanth-lingappa) [inline python_builtin_functions.py]: Again, why converting `x` to `int`

File: python_builtin_functions.py [hunk 15]
```diff
@@ -4,86 +4,170 @@ def myabs(number):
...
+        lst.insert(1,str(r))
+    return ''.join(lst)
 
 
 def myisinstance(object, classinfo):
     """ Implementation of built-in `isinstance` function """
-    pass
+    t1=type(object)
+    t2=str(t1).find(classinfo)
+    if t2>=0:
```
Comment (sumanth-lingappa) [inline python_builtin_functions.py]: Though this works, this is a bit confusing..
Try as below..

```python
if str(classinfo) in type(object):
    return True
return False
```

File: python_builtin_functions.py [hunk 16]
```diff
@@ -4,86 +4,170 @@ def myabs(number):
...
+    dot='.'
+    lst=list(str(number))
+    f=str(number).find(dot)
+    t=f+ndigits+1
+    if int(lst[t])>5:
+        lst[t-1]=str(int(lst[t-1])+1)
+    return  ''.join(lst[0:t])
 
 
 def mysorted(iterable, reverse=False):
```
Comment (sumanth-lingappa) [inline python_builtin_functions.py:168]: Can you name the `sorting technique` used here?

File: python_builtin_functions.py [hunk 17]
```diff
@@ -4,86 +4,170 @@ def myabs(number):
...
     """ Implementation of built-in `isinstance` function """
-    pass
+    t1=type(object)
+    t2=str(t1).find(classinfo)
+    if t2>=0:
+        return True
+    return False
 
 
 def myround(number, ndigits=None):
```
Comment (sumanth-lingappa) [inline python_builtin_functions.py:157]: Check if this works for 
- `myround(3.5)`
- `myround(2.5)`

**Review by sumanth-lingappa — CHANGES_REQUESTED** [review]

File: python_builtin_functions.py (hunk 1 above)
Comment (VikramKuberappa) [reply python_builtin_functions.py]: Changed.

File: python_builtin_functions.py (hunk 2 above)
Comment (VikramKuberappa) [reply python_builtin_functions.py]: Given proper naming.

File: python_builtin_functions.py (hunk 3 above)
Comment (VikramKuberappa) [reply python_builtin_functions.py]: removed else block

File: python_builtin_functions.py (hunk 4 above)
Comment (VikramKuberappa) [reply python_builtin_functions.py]: Changed as per your instruction.

File: python_builtin_functions.py (hunk 5 above)
Comment (VikramKuberappa) [reply python_builtin_functions.py]: Changed as per your instruction.

File: python_builtin_functions.py (hunk 13 above)
Comment (VikramKuberappa) [reply python_builtin_functions.py:133]: Changed code and its working.

File: python_builtin_functions.py (hunk 14 above)
Comment (VikramKuberappa) [reply python_builtin_functions.py]: removed int.

File: python_builtin_functions.py (hunk 15 above)
Comment (VikramKuberappa) [reply python_builtin_functions.py]: Changed as per your instruction.

File: python_builtin_functions.py (hunk 16 above)
Comment (VikramKuberappa) [reply python_builtin_functions.py:168]: Bubble sorting technique

File: python_builtin_functions.py (hunk 17 above)
Comment (VikramKuberappa) [reply python_builtin_functions.py:157]: Now i changed the code and its working fine for the above mentioned input.

Comment (sumanth-lingappa) [reply python_builtin_functions.py:157]: `round()` is a bit complicated to implement.

For now we will understand how it works, and not implement.

**Pull request closed without merging**